### ISSUE TITLE
Wrong IC_SUPERCLASS_USES_SUBCLASS_DURING_INITIALIZATION Warning

### ISSUE THREAD

**Issue by gohrner** [spotbugs/spotbugs#1360]
For the following class, SpotBugs reports IC_SUPERCLASS_USES_SUBCLASS_DURING_INITIALIZATION:

```java
package ghIssues;

public interface Issue1234 {
	public static final Issue1234 REF_TO_INNER = InnerClass.INSTANCE; 

	public static final class InnerClass implements Issue1234 {
		public static final InnerClass INSTANCE = new InnerClass();
	}
}
```

Independently of coding style and whether this is a good idea or confusing, accoring to JLS section 12.4 - if I understand it correctly - this should be fine: Initialization of the inner class does not cause the implemented interface to be initialized.

So once interface `Issue1234` starts initialization, the assignment of `REF_TO_INNER` triggers initialization of `InnerClass`, which ensures that `INSTANCE` is defined before assignment.

Or did I get something wrong?

The SpotBugs warning's explanation is:

> Bug: Initialization of ghIssues.Issue1234 accesses class ghIssues.Issue1234$InnerClass, which isn't initialized yet
During the initialization of a class, the class makes an active use of a subclass. That subclass will not yet be initialized at the time of this use. For example, in the following code, foo will be null.
>
> ```java
public class CircularClassInitialization {
    static class InnerClassSingleton extends CircularClassInitialization {
        static InnerClassSingleton singleton = new InnerClassSingleton();
    }

    static CircularClassInitialization foo = InnerClassSingleton.singleton;
}```

This has two nested classes - as opposed to an interface containing a subclass as an inner class - and in this case this actually might be wrong: The JLS states that a class' initialization triggers the initialization of it superclass, and while more recent JLS' don't explicitly mention the order in which sub- and superclass are initialized, the examples there clearly state that superclasses are initialized first. JLS for Java 7 was more explicit in this regard and clearly stated, that a superclass is initialized before the subclass, in which case the code above would probably assign `null` to `REF_TO_INNER`.

(Note: The attached files all reference an example issue number, as I don't yet know which number this issue will get assigned after submitting it...)

jUnit Test class:

```java
/*
 * Contributions to SpotBugs
 * Copyright (C) 2020, gohrner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
package edu.umd.cs.findbugs.ba;

import static org.hamcrest.Matchers.hasSize;
import static org.junit.Assert.assertThat;

import org.junit.Test;

import edu.umd.cs.findbugs.AbstractIntegrationTest;
import edu.umd.cs.findbugs.SortedBugCollection;
import ghIssues.Issue1234;

public class Issue1234Test extends AbstractIntegrationTest {

    @Test
    public void test() {
        performAnalysis("ghIssues/Issue1234.class", "ghIssues/Issue1234$InnerClass.class");
        SortedBugCollection bugCollection = (SortedBugCollection) getBugCollection();
        
        assertThat(bugCollection.getCollection(), hasSize(0));
    }
}
```

**Comment by ThrawnCA** [comment]
So, if I'm understanding correctly, you believe that this bug detector should not fire when the enclosing class is actually an interface?

**Comment by gohrner** [comment]
Yes, I think this would be more correct. To be precise, according to the current JLS, if the enclosing class is an interface without default methods.

BTW, I could not find any case with the Oracle Java / OpenJDK versions I tried where the example code from the SpotBugs bug explanation actually produces "null", even if I used an interface with default methods. This however might be implementation specific, of course.

Maybe there's someone reading here who could shed more light on this?

**Comment by gohrner** [comment]
Another question is if the SpotBugs code example given for the `IC_SUPERCLASS_USES_SUBCLASS_DURING_INITIALIZATION` bug is actually correct - can/will the field in this example even be `null`? Or will a value to an actual object instance, which however might only be partially initialized or might be uninitialized, be assigned in any case?

```java
public class CircularClassInitialization {
	static class InnerClassSingleton extends CircularClassInitialization {
		static InnerClassSingleton singleton = new InnerClassSingleton();
	}
	
	static CircularClassInitialization foo = InnerClassSingleton.singleton;
}
```

Or does the "final" in my example make a difference? (It's not included in the example snipped output by SpotBugs.)

I don't seem to be the first who stumbled across this detector:

* https://mailman.cs.umd.edu/pipermail/findbugs-discuss/2007-March/001832.html
* https://mailman.cs.umd.edu/pipermail/findbugs-discuss/2012-July/003601.html
